GDAXSharp/WebSocket/WebSocket.cs

```diff
@@ -0,0 +1,212 @@
+using GDAXSharp.WebSocket.Types;
+using SuperSocket.ClientEngine;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Net.Http;
+using WebSocket4Net;
+
+namespace GDAXSharp.WebSocket
+{
+    public class WebSocket : AbstractRequest, IWebSocket
+    {
+        private const string ApiUri = "wss://ws-feed.gdax.com";
+
+        private const string SandBoxApiUri = "wss://ws-feed-public.sandbox.gdax.com";
+
+        internal WebSocket(
+            IAuthenticator authenticator,
+            IClock clock,
+            bool sandBox = false)
+                : base(authenticator, clock, sandBox)
+        {
+        }
+
+        ~WebSocket()
+        {
+            WebSocketFeed.Close();
+            WebSocketFeed.Dispose();
+        }
+
+        private bool StopWebSocket { get; set; }
+
+        private List<ProductType> ProductTypes { get; set; }
+
+        private List<ChannelType> ChannelTypes { get; set; }
+
+        private WebSocket4Net.WebSocket WebSocketFeed { get; set; }
+
+        public void Start(List<ProductType> productTypes, List<ChannelType> channelTypes = null)
+        {
+            if (WebSocketFeed != null && WebSocketFeed.State != WebSocketState.Closed)
+            {
+                throw new GDAXSharpHttpException(
+                    $"Websocket needs to be in the closed state, current state is {WebSocketFeed.State}");
+            }
+
+            if (productTypes.Count == 0)
+            {
+                throw new ArgumentException("You must specify at least one product type");
+            }
+
+            StopWebSocket = false;
+
+            ProductTypes = productTypes;
+            ChannelTypes = channelTypes;
+
+            var socketUrl = SandBox
+                ? SandBoxApiUri
+                : ApiUri;
+
+            WebSocketFeed = new WebSocket4Net.WebSocket(socketUrl);
+            WebSocketFeed.Closed += WebSocket_Closed;
+            WebSocketFeed.Error += WebSocket_Error;
+            WebSocketFeed.MessageReceived += WebSocket_MessageReceived;
+            WebSocketFeed.Opened += WebSocket_Opened;
+            WebSocketFeed.Open();
+        }
+
+        public void Stop()
+        {
+            if (WebSocketFeed == null || WebSocketFeed.State != WebSocketState.Open)
+            {
+                throw new GDAXSharpWebSocketException($"Websocket needs to be in the opened state, current state is {WebSocketFeed?.State ?? WebSocketState.None}")
+                {
+                    WebSocket = WebSocketFeed,
+                    StatusCode = WebSocketFeed?.State ?? WebSocketState.None,
+                    ErrorEvent = null
+                };
+            }
+
+            if (StopWebSocket)
+            {
+                throw new GDAXSharpWebSocketException("Websocket can't be stopped")
+                {
+                    WebSocket = WebSocketFeed,
+                    StatusCode = WebSocketFeed?.State ?? WebSocketState.None,
```

Comment:
you can remove the right side of the `??` because the left side will never be null